File: jinia/object/src/main/java/chapter3/ReservationAgency.java
package chapter3;

public class ReservationAgency {
    // old
//    public Reservation reserve(Customer customer, Screening screening, int audienceCount) {
//        Movie movie = screening.getMovie();
//        // 해당하는 정책 찾기
//        boolean discountable = false;
//        for (DiscountCondition discountCondition : movie.getDiscountConditions()) {
//            if (discountCondition.getType() == DiscountConditionType.PERIOD) {
//                discountable = screening.getWhenScreened().getDayOfWeek().equals(discountCondition.getDayOfWeek()) &&
//                        discountCondition.getStartTime().compareTo(screening.getWhenScreened().toLocalTime()) <= 0 &&
//                        discountCondition.getEndTime().compareTo(screening.getWhenScreened().toLocalTime()) >= 0;
//            } else {
//                discountable = discountCondition.getSequence() == screening.getSequence();
//            }
//            if(discountable) break;
//        }
//
//        // 정책으로 요금 정하기
//        Money fee;
//
//        if(discountable) {
//            Money discountAmount = null;
//            switch (movie.getMovieType()) {
//                case AMOUNT_DISCOUNT:
//                    discountAmount = movie.getDiscountAmount();
//                    break;
//                case PERCENT_DISCOUNT:
//                    discountAmount = movie.getFee().times(movie.getDiscountPercent());
//                    break;
//                case NONE_DISCOUNT:
//                    discountAmount = Money.ZERO;
//                    break;
//            }
//            fee = movie.getFee().minus(discountAmount);
//        }else {
//            fee = movie.getFee();
//        }
//        return new Reservation(customer, screening, fee, audienceCount);
//    }

    // refactoring
    public Reservation reserve(Customer customer, Screening screening, int audienceCount){
        Money fee = screening.calculateFee(audienceCount);
        return new Reservation(customer, screening, fee, audienceCount);
    }
}

File: prravda/src/data-driven-ticket-selling-application/reservation-agency.ts
import { Screening } from "./screening";
import { Money } from "../theater-booking-application/money";
import { AmountDiscountPolicy } from "../theater-booking-application/discount-policies/amount-discount-policy";
import { MovieType } from "./movie-type";

export class ReservationAgency {
  constructor(
    private screening: Screening,
    private customer: Customer,
    private audienceCount: number
    private reservation: Reservation,
  ) {
    let discountable: boolean = false;
    const discountConditionList = screening.movie.discountConditions;
    for (const eachCondition of discountConditionList) {
      if (discountable) {
        break;
      }
      if (eachCondition.type === "PERIOD") {
        discountable = Boolean(
          screening.whenScreened.getDay() === eachCondition.dayOfWeek &&
            eachCondition.startTime.getTime() >
              screening.whenScreened.getTime() &&
            eachCondition.endTime.getTime() < screening.whenScreened.getTime()
        );
      }
      discountable = Boolean(eachCondition.sequence === screening.sequence);

      let fee: Money;
      if (discountable) {
        let discountAmount: Money = Money.ZERO;
        switch (screening.movie) {
          case MovieType.AMOUNT_DISCOUNT:
            discountAmount = screening.movie.discountAmount;
            break;
          case MovieType.PERCENT_DISCOUNT:
            discountAmount = screening.movie.discountPercent;
            break;
            case MovieType.NONE_DISCOUNT:
            default:
            discountAmount = screening.movie.;
            break;
        }
          fee = screening.movie.fee.minus(dicountAmount);
      } else {
          fee = screening.movie.fee;
      }
      this.reservation = new Reservation(this.customer, this.screening, fee, audienceCount);
    }
  }
}
